combining witness problem with ground checker

qcorporation opened the pull request:
adding test cases
removing visitNegation call within AstGroundAnalysis

Commit:
combining witness problem with ground checker
adding test cases
removing visitNegation call within AstGroundAnalysis

ohamel-softwaresecure commented on src/AstSemanticChecker.cpp:
```diff
@@ -1015,15 +1013,13 @@ void AstSemanticCheckerImpl::checkIO() {
-        const std::set<std::string>& getAggregatorVariables() {
-            return aggregatorVariables;
+        mutable std::set<std::string> variables;
+        const std::set<std::string>& getVariables() {
```
why bother? `variables` is public.

qcorporation replied:
remove getter

ohamel-softwaresecure commented on src/AstSemanticChecker.cpp:
```diff
@@ -1032,26 +1028,27 @@ static const std::vector<SrcLocation> usesInvalidWitness(AstTranslationUnit& tu,
 
                 return std::make_unique<AstVariable>(newVariableName.str());
+            } else if (dynamic_cast<AstNegation*>(node.get())) {
+                // replace all negation terms with a boolean constraint of 'true'
```
Comment should explain *why* (and maybe *how*, if tricky) not *what*.

qcorporation replied:
updated comments

ohamel-softwaresecure commented on src/AstSemanticChecker.cpp:
```diff
@@ -1032,26 +1028,27 @@ static const std::vector<SrcLocation> usesInvalidWitness(AstTranslationUnit& tu,
     // Clause 2 - will have aggregators replaced with intrinsically grounded variables
-    auto aggregatorlessClause = std::make_unique<AstClause>();
-    aggregatorlessClause->setHead(std::make_unique<AstAtom>("*"));
+    auto lessClause = std::make_unique<AstClause>();
```
Rename `erasedClause`?

qcorporation replied:
renamed to `erasedClause`

ohamel-softwaresecure commented on src/AstSemanticChecker.cpp:
```diff
@@ -1524,6 +1496,26 @@ void GroundedTermsChecker::verify(AstTranslationUnit& translationUnit) {
+        std::set<std::unique_ptr<AstArgument>> groundedArguments;
+        for (auto&& invalidArgument : usesInvalidWitness(translationUnit, bodyLiterals, groundedArguments)) {
+            report.addError(
+                    "Witness problem: argument grounded by an aggregator/negation's inner scope is used "
+                    "ungrounded in outer scope",
+                    invalidArgument);
```
grammar

qcorporation replied:
updated text

ohamel-softwaresecure commented on tests/evaluation/neg8/neg8.dl:
```diff
@@ -0,0 +1,17 @@
+.decl b(x: number)
```
Add "souffle devs 2020" header and a very short blurb explaining what behaviour this test case exercises.

qcorporation replied:
added header

ohamel-softwaresecure reviewed (changes requested):
generally looks good. I'd've loved if the duplication work in check-witness could have been reduced, but that's not critical.

Commit:
addressing comments within PR
- grammar
- eraseClause renaming
- header for test cases
- removal of public getter

Commit:
Merge branch 'task/qcorp/OM-1260' into task/qcorp/OM-1260-neg-single-use-var

Commit:
Merge branch 'task/qcorp/OM-1260' into task/qcorp/OM-1260-neg-single-use-var

Commit:
fixing groundnessConstraint going out of scope and getting freed

Commit:
simplify `usesInvalidWitness`

ohamel-softwaresecure commented:
LGTM
